core/src/test/java/org/apache/accumulo/core/lock/ServiceLockPathsTest.java

```diff
@@ -0,0 +1,880 @@
+    assertEquals(1, results.size());
+    iter = results.iterator();
+    slp1 = iter.next();
+    assertEquals(HOSTNAME, slp1.getServer());
+    assertEquals(ZCOMPACTORS, slp1.getType());
+    assertEquals(TEST_RESOURCE_GROUP, slp1.getResourceGroup());
+    assertEquals(ROOT + ZCOMPACTORS + "/" + TEST_RESOURCE_GROUP + "/" + HOSTNAME, slp1.toString());
+
+    // query for a wrong server
+    results = ctx.getServerPaths().getCompactor(Optional.of(TEST_RESOURCE_GROUP),
+        Optional.of(HostAndPort.fromString("localhost:1234")));
+    assertEquals(0, results.size());
+
+    EasyMock.verify(ctx, zc);
+
+  }
+
+  @Test
+  public void testGetScanServersNotRunning() {
+
+    ClientContext ctx = EasyMock.createMock(ClientContext.class);
+    ZooCache zc = EasyMock.createMock(ZooCache.class);
+
+    EasyMock.expect(ctx.getZooKeeperRoot()).andReturn(ROOT).anyTimes();
+    EasyMock.expect(ctx.getZooCache()).andReturn(zc).anyTimes();
+    EasyMock.expect(zc.getChildren(ROOT + ZSSERVERS)).andReturn(List.of()).anyTimes();
+    EasyMock.expect(ctx.getServerPaths()).andReturn(new ServiceLockPaths(ctx)).anyTimes();
+    EasyMock.replay(ctx, zc);
+
+    assertThrows(NullPointerException.class, () -> ctx.getServerPaths().getScanServer(null, null));
+    assertThrows(NullPointerException.class,
+        () -> ctx.getServerPaths().getScanServer(Optional.of(TEST_RESOURCE_GROUP), null));
+    assertTrue(ctx.getServerPaths().getScanServer(Optional.empty(), Optional.empty()).isEmpty());
+    assertTrue(ctx.getServerPaths()
+        .getScanServer(Optional.of(TEST_RESOURCE_GROUP), Optional.empty()).isEmpty());
+    assertTrue(ctx.getServerPaths().getScanServer(Optional.of(TEST_RESOURCE_GROUP), Optional.of(hp))
+        .isEmpty());
+
+    EasyMock.verify(ctx, zc);
+
+  }
+
+  @Test
+  public void testGetScanServers() {
+
+    UUID uuid = UUID.randomUUID();
+    String svcLock1 = ServiceLock.ZLOCK_PREFIX + uuid.toString() + "#0000000001";
+    String svcLock2 = ServiceLock.ZLOCK_PREFIX + uuid.toString() + "#0000000002";
+    ServiceLockData sld1 =
+        new ServiceLockData(uuid, HOSTNAME, ThriftService.TABLET_SCAN, TEST_RESOURCE_GROUP);
+    ServiceLockData sld2 =
+        new ServiceLockData(uuid, HOSTNAME, ThriftService.TABLET_SCAN, DEFAULT_RESOURCE_GROUP_NAME);
+
+    ClientContext ctx = EasyMock.createMock(ClientContext.class);
+    ZooCache zc = EasyMock.createMock(ZooCache.class);
+
+    EasyMock.expect(ctx.getZooKeeperRoot()).andReturn(ROOT).anyTimes();
+    EasyMock.expect(ctx.getZooCache()).andReturn(zc).anyTimes();
+    EasyMock.expect(zc.getChildren(ROOT + ZSSERVERS))
+        .andReturn(List.of(TEST_RESOURCE_GROUP, DEFAULT_RESOURCE_GROUP_NAME)).anyTimes();
+    EasyMock.expect(zc.getChildren(ROOT + ZSSERVERS + "/" + TEST_RESOURCE_GROUP))
+        .andReturn(List.of(HOSTNAME)).anyTimes();
+    EasyMock.expect(zc.getChildren(ROOT + ZSSERVERS + "/" + DEFAULT_RESOURCE_GROUP_NAME))
```

Comment:
Would it be worthwhile to test more than one sever in a RG and zero servers in an RG in ZK?

Reply:
I think that ServiceLockPathsIT tests this case.